### PR TITLE
[Add] `IEquatable` to WitnessRule

### PR DESCRIPTION
**Pull request by cschuchardt88**
# Change Log

- Added `IEquatable` to `WitnessRule` class
- Added Unit Tests for `IEquatable` for `WitnessRule` class

Needs #3572

## Type of change

- [ ] Optimization (the change is only an optimization)
- [ ] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Test_IEquatable

**Test Configuration**:
- Unit Tests

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/Neo/Network/P2P/Payloads/WitnessRule.cs [hunk 1]
```diff
@@ -37,6 +37,27 @@ public class WitnessRule : IInteroperable, ISerializable
 
         int ISerializable.Size => sizeof(WitnessRuleAction) + Condition.Size;
 
+        public bool Equals(WitnessRule other)
+        {
```
Comment (shargon) [inline src/Neo/Network/P2P/Payloads/WitnessRule.cs:41]: It fail with null

Comment (cschuchardt88) [reply src/Neo/Network/P2P/Payloads/WitnessRule.cs:41]: what do you mean? exception? `WitnessRule` is not `nullable`.

Comment (shargon) [reply src/Neo/Network/P2P/Payloads/WitnessRule.cs:41]: Because we haven't nullable enabled, also is not in object and is checked

Comment (cschuchardt88) [reply src/Neo/Network/P2P/Payloads/WitnessRule.cs:41]: ok, fixed

File: src/Neo/Network/P2P/Payloads/WitnessRule.cs [hunk 2]
```diff
@@ -96,5 +118,23 @@ public StackItem ToStackItem(IReferenceCounter referenceCounter)
...
+            return left.Equals(right);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool operator !=(WitnessRule left, WitnessRule right)
+        {
+            if (left is null || right is null)
+                return !Equals(left, right);
+
+            return !(left.Equals(right));
```
Comment (cschuchardt88) [inline src/Neo/Network/P2P/Payloads/WitnessRule.cs:137]:
```suggestion
            return !left.Equals(right);
```

**Review by Jim8y — APPROVED** [review]

**Review by nan01ab — APPROVED** [review]

**Review by shargon — APPROVED** [review]

**Pull request merged**